source/slang/source-loc.h

```diff
@@ -3,16 +3,28 @@
 #define SLANG_SOURCE_LOC_H_INCLUDED
 
 #include "../core/basic.h"
-#include "../../slang-com-ptr.h"
+#include "../core/slang-memory-arena.h"
+#include "../core/slang-string-slice-pool.h"
 
+#include "../../slang-com-ptr.h"
 #include "../../slang.h"
 
 namespace Slang {
 
+/** Overview: 
+
+SourceFile - Is the immutable contents of a file (or perhaps some generated source - say from doing a macro substitution)
+SourceUnit - Tracks a single parse of a SourceFile. Each SourceUnit defines a range of source locations used. If a SourceFile is parsed twice, two 
+SourceUnits are used, with unique SourceRanges - such that it is possible to tell which specific parse a SourceLoc is from. Not only that but the SourceUnit 
+contains the modifications of the interpretations of source (say by #line) directives. It is necessary to have these different 'views' on the 
+source because if a file is included twice, it may be used in different ways and have different #line directives.
+
+*/
+
 class SourceLoc
 {
 public:
-    typedef UInt RawValue;
+    typedef uint32_t RawValue;
```

Comment:
If we are going to make this fixed-size then I might prefer 64-bit as the default.

Because the compiler keeps all source files resident in memory for the duration of a compile, though, there is little argument in favor of making this be larger than the size of a pointer (if you really need 64 bits of source locations on a 32-bit build, you must be doing a lot of repeated `#include`s).

Reply:
We'd prefer 32 bits over 64 - for space saving and better cache coherency (memory consumption dropped > 2Mb just from this change). It was also preferable for serializability and remapping. 

That said these are not good enough reasons if it means people can't compile real programs. I think in practice though we are very safe here. With the change we went from using towards 30+ bits. Now we are around 19 bits, and thats with > 600k of source. We have 13 bits of room - if anybody comes close to such a limit anytime soon they are writing some pretty huge shaders, or doing something pretty odd. 

But if it really shown that that many bits are needed in the real world even very very rarely it's an easy change.